cudax/include/cuda/experimental/__container/async_mdarray.cuh

```diff
@@ -0,0 +1,901 @@
+  //! that holds the memory resource used to allocate the async_mdarray
+  //! @endrst
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI const __resource_t& get_memory_resource() const noexcept
+  {
+    return __buf_.get_memory_resource();
+  }
+
+  //! @brief Returns the stored stream
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI constexpr ::cuda::stream_ref get_stream() const noexcept
+  {
+    return __buf_.get_stream();
+  }
+
+  //! @brief Replaces the stored stream
+  //! @param __new_stream the new stream
+  //! @note Always synchronizes with the old stream
+  _CCCL_HIDE_FROM_ABI constexpr void change_stream(::cuda::stream_ref __new_stream)
+  {
+    __buf_.change_stream(__new_stream);
+  }
+
+  //! @brief Replaces the stored stream
+  //! @param __new_stream the new stream
+  //! @warning This does not synchronize between \p __new_stream and the current stream. It is the user's responsibility
+  //! to ensure proper stream order going forward
+  _CCCL_HIDE_FROM_ABI constexpr void change_stream_unsynchronized(::cuda::stream_ref __new_stream) noexcept
+  {
+    __buf_.change_stream_unsynchronized(__new_stream);
+  }
+
+  //! @brief Returns the execution policy
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI constexpr __policy_t get_execution_policy() const noexcept
+  {
+    return __policy_;
+  }
+
+  //! @brief Replaces the currently used execution policy
+  //! @param __new_policy the new policy
+  _CCCL_HIDE_FROM_ABI constexpr void set_execution_policy(__policy_t __new_policy) noexcept
+  {
+    __policy_ = __new_policy;
+  }
+
+  //! @brief Returns the execution policy
+  _CCCL_HIDE_FROM_ABI void wait() const
+  {
+    __buf_.get_stream().wait();
+  }
+  //! @}
+
+  //! @brief Swaps the contents of a async_mdarray with those of \p __other
+  //! @param __other The other async_mdarray.
+  _CCCL_HIDE_FROM_ABI void swap(async_mdarray& __other) noexcept
+  {
+    _CUDA_VSTD::swap(__buf_, __other.__buf_);
+    _CUDA_VSTD::swap(size(), __other.size());
+  }
+
+  //! @brief Swaps the contents of two async_vectors
+  //! @param __lhs One async_mdarray.
+  //! @param __rhs The other async_mdarray.
+  _CCCL_HIDE_FROM_ABI friend void swap(async_mdarray& __lhs, async_mdarray& __rhs) noexcept
+  {
+    __lhs.swap(__rhs);
+  }
+  //! @}
+
+  //! @addtogroup comparison
+  //! @{
+
+  //! @brief Compares two async_vectors for equality
+  //! @param __lhs One async_mdarray.
+  //! @param __rhs The other async_mdarray.
+  //! @return true, if \p __lhs and \p __rhs contain equal elements have the same size
+  _CCCL_NODISCARD_FRIEND _CCCL_HIDE_FROM_ABI constexpr bool
+  operator==(const async_mdarray& __lhs, const async_mdarray& __rhs) noexcept(noexcept(_CUDA_VSTD::equal(
+    __lhs.__unwrapped_begin(), __lhs.__unwrapped_end(), __rhs.__unwrapped_begin(), __rhs.__unwrapped_end())))
+  {
+    ::cuda::experimental::stream_ref{__lhs.get_stream()}.wait(__rhs.get_stream());
+    return __lhs.__equality(
```

Comment:
This is only correct if the layout mapping is exhaustive.  It's only efficient if the layout mapping is unique.

Reply:
dropped